### ISSUE TITLE
Thumbnailed images generate incorrect HTML and/or Markdown

### ISSUE THREAD

**Issue by eadmund** [kaushalmodi/ox-hugo#125]
`[[file:images/foo.jpeg][file:images/bar.jpeg]]` generates `[{{<figure src="/images/bar.jpeg">}}](/images/foo.jpeg)`, which turns into:

```
<p><a href="/images/foo.jpeg">
<figure >
    
        <img src="/images/bar.jpeg" />
    
    
</figure>
</a></p>
```

This HTML doesn't pass validation because the `p` & `a` tags should be closed before the `figure` tag, and the `/a` & `/p` tags close already-closed tags.

It would be better if thumbnailed image blocks (as opposed to images inline in text) were recognised, and instead generated this Markdown: `{{<figure src="/images/bar.jpeg" link="/images/foo.jpeg">}}`.

Also, inline images generate incorrect Markdown: `Some text [[file:images/foo.jpeg][file:images/bar.jpeg]]` turns into `Some text [{{<figure src="/images/bar.jpeg">}}](/images/bar.jpeg)`, which ends up generating the incorrect HTML:

```
<p>Some text: <a href="/images/foo.jpeg">
<figure >
    
        <img src="/images/bar.jpeg" />
    
    
</figure>
</a></p>
```

It should instead be `Some text [![](/images/bar.jpeg)](/images/foo.jpeg)` (alt text could go between the `[]`).

To sum up:

| Input | Has link? | Inline? | Output |
|----|----|----|----|
| `[[file:/images/foo.jpeg][file:/images/bar.jpeg]]` | `t` | `nil` | `{{figure src="/images/bar.jpeg" link="/images/foo.jpeg"}}` |
| `[[file:/images/foo.jpeg]]` | `nil` | `nil` | `{{figure src="/images/bar.jpeg"}}` |
| `Some text [[file:/images/foo.jpeg][file:/images/bar.jpeg]] more text` | `t` | `t` | `Some text [![](/images/bar.jpeg)](/images/foo.jpeg)` |
| `Some text [[file:/images/foo.jpeg]]` | `nil` | `t` | ` Some text ![](/images/foo.jpeg)` |

N.b.: Markdown does not offer a way to set HTML attributes other than `alt` (e.g. `height`, `width` or `class`) on image references, so if those are provided with `ATTR_HTML` then an HTML `img` tag should be used instead.

**Comment by kaushalmodi** [comment]
Thanks for the detailed summary. I was not aware of these issues as I don't know much of HTML. 

I will start tackling these issues one at a time, it will take time.

- [ ] Non-inlined hyperlinked images -- Use `link` in `figure`
- [ ] Inlined non-hyperlinked images, no meta or only `alt` -- Use Markdown image syntax
- [ ] Inlined hyperlinked images, no meta or only `alt` -- Use Markdown image syntax
- [ ] Inlined non-hyperlinked images, meta other than`alt` -- Use HTML `img`
- [ ] Inlined hyperlinked images, meta other than `alt` -- Use HTML `img`

**Comment by kaushalmodi** [comment]
@eadmund All of the issues you mentioned are now fixed. Please try it out now.

PS: I allowed one regression in the fix in https://github.com/kaushalmodi/ox-hugo/commit/e92fcf00375e9b20b4e9029b03fdc16409bd6f30: Now `target`, `rel`, etc. `<a>` tag parameters set in `#+ATTR_HTML` will get discarded **only for** non-inlined, hyperlinked image (because of `figure` shortcode limitation). It didn't seem a common use case that deemed a fix at this time.

**Comment by kaushalmodi** [comment]
If interested in having the `target` and `rel` parameters to the internal `figure` shortcode, give a shout of approval [here, on the Hugo Discourse forum][1].

[1]: https://discourse.gohugo.io/t/proposal-to-add-target-and-ref-parameters-to-the-figure-shortcode/10300

**Comment by eadmund** [comment]
One thing you might want to consider is not using the `figure` shortcode at all — it's a convenient shorthand for human writers, but ~ox-hugo~ could just write out the `figure` HTML tags, since HTML is legal in Markdown.

**Comment by kaushalmodi** [comment]
One of the reasons for ox-hugo to export to Markdown is also to enable collaboration on Hugo content with non-Org users. Otherwise, I could have simply made ox-hugo export *everything* in HTML, with front-matter attached at the top :)

Personally, I find the `figure` shortcode a bit more readable than the `figure` HTML code (as I don't have a raw HTML coding experience). 

For this particular use case, I'm a bit optimistic that the `target` might get into the official shortcode. And if not, I'll publish a custom shortcode for folks to add to their site if they want the `target` and `rel` attrs inside `figure`. I think that would be much easier than fully supporting figure in HTML. 

*PS: I am contributing to official Hugo documentation by ox-hugo exported Markdown.*

**Comment by eadmund** [comment]
Ah, I didn't realise that you wanted to support collaboration with non-Org users. That makes sense, although it _might_ still be worth using the `figure` tag in some cases.

Thanks for all the work on `ox-hugo` BTW. I'm opening all these issues because I think it's awesome and am hoping to help make it better:-)

**Comment by kaushalmodi** [comment]
Thanks. I actually appreciate these issues opened  by you and the feedback I get. Keep them coming!

Btw, let me know if the original issue of this thread got resolved. If so, I'll close this issue. 

I'll track the `target` support in `figure` in a new issue.

**Comment by eadmund** [comment]
It seems to have done the trick for my use case!
